src/bucket/test/BucketIndexTests.cpp

```diff
@@ -272,6 +357,177 @@ class BucketIndexTest
+        UnorderedSet<LedgerKey> notLoaded;
+        LedgerKeySet expectedSuccessKeys;
+        LedgerKeySet expectedFailedKeys;
+        UnorderedMap<LedgerKey, std::uint32_t> keyToEntrySize;
+        std::vector<Hash> txs(4, Hash());
+        for (uint8_t i = 0; i < txs.size(); i++)
+        {
+            txs[i][0] = i;
+        }
+
+        expectedSuccessKeys.insert(insertContractDataKeyMinSize(
+            page_size, txs[0], lkToTx, txReadBytes, keyToEntrySize));
+        // loadKeys will load an entry if it has a transaction with a non-zero
+        // quota. Even with a quota of 1, we can load an entry with a size
+        // larger than the page size.
+        expectedSuccessKeys.insert(insertContractDataKeyMinSize(
+            page_size, txs[1], lkToTx, txReadBytes, keyToEntrySize));
+        txReadBytes[txs[1]] = 1;
+        // tx[2] has zero quota, so no keys should succeed.
+        expectedFailedKeys.insert(insertContractDataKeyMinSize(
+            page_size, txs[2], lkToTx, txReadBytes, keyToEntrySize));
+        txReadBytes[txs[2]] = 0;
+
+        auto key_a = insertContractDataKeyMinSize(page_size, txs[3], lkToTx,
+                                                  txReadBytes, keyToEntrySize);
+        auto key_b = insertContractDataKeyMinSize(page_size, txs[3], lkToTx,
+                                                  txReadBytes, keyToEntrySize);
+        txReadBytes[txs[3]] -=
+            std::max(keyToEntrySize[key_a], keyToEntrySize[key_b]);
+
+        auto loadResult = getBM().loadKeysWithLimits(mKeysToSearch, lkToTx,
+                                                     txReadBytes, notLoaded);
+        LedgerKeySet loadResultSet{};
+        std::transform(
+            loadResult.begin(), loadResult.end(),
+            std::inserter(loadResultSet, loadResultSet.end()), [](auto& le) {
+                return LedgerEntryKey(const_cast<const LedgerEntry&>(le));
+            });
+        // Either a or b should be loaded, but not both.
+        if (notLoaded.find(key_a) == notLoaded.end())
+        {
+            expectedSuccessKeys.insert(key_a);
+            expectedFailedKeys.insert(key_b);
+        }
+        else
+        {
+            expectedFailedKeys.insert(key_a);
+            expectedSuccessKeys.insert(key_b);
+        }
+        REQUIRE(loadResult.size() == expectedSuccessKeys.size());
+        REQUIRE(notLoaded.size() == expectedFailedKeys.size());
+
+        for (auto key : expectedSuccessKeys)
+        {
+            REQUIRE(notLoaded.find(key) == notLoaded.end());
+            REQUIRE(loadResultSet.find(key) != loadResultSet.end());
+        }
+        for (auto key : expectedFailedKeys)
+        {
+            REQUIRE(notLoaded.find(key) != notLoaded.end());
+            REQUIRE(loadResultSet.find(key) == loadResultSet.end());
+        }
+        for (auto tx : txs)
+        {
+            REQUIRE(txReadBytes[tx] == 0);
+        }
+    }
+
+    void
+    testLoadContractKeysWithInsufficientReadBytes()
```

Comment:
This test is pretty hard to follow, I think it would be easier if we chatted and you walked me through it.